internal/plugin/metrics.go

```diff
@@ -0,0 +1,303 @@
+/**
+ * Copyright (c) F5, Inc.
+ *
+ * This source code is licensed under the Apache License, Version 2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"log/slog"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/nginx/agent/v3/internal/bus"
+	"github.com/nginx/agent/v3/internal/config"
+	"github.com/nginx/agent/v3/internal/metrics"
+	"github.com/nginx/agent/v3/internal/metrics/prometheus"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+
+	metricSdk "go.opentelemetry.io/otel/sdk/metric"
+)
+
+const otelServiceName = "nginx-agent"
+
+type (
+	shutdownFunc func()
+
+	DataSource interface {
+		// Starts the data source and begins data collection at a regular interval. Metrics should be pushed to the
+		// MessagePipe using the metrics topic, as they will be handled in the metrics plugin.
+		Start(ctx context.Context, updateInterval time.Duration) error
+		// Stops the data source from collecting data.
+		Stop()
+		// Getter for the type of data source.
+		Type() string
+	}
```

Comment:
I would also like to potentially change this contract, as it does not clearly define how a data source should behave (in this case, push metrics to the metrics topic at the configured update interval). I also want to make sure that having a push model is better than a pull model for our use cases.